src/modules/frost/session_impl.h

```diff
@@ -0,0 +1,646 @@
+        secp256k1_scalar sk;
+        int ret_tmp;
+
+        ret &= secp256k1_frost_share_load(ctx, &sk, agg_share);
+        secp256k1_scalar_clear(&sk);
+
+        ret_tmp = secp256k1_frost_share_serialize(ctx, sk_ser, agg_share);
+        VERIFY_CHECK(ret_tmp);
+        sk_ser_ptr = sk_ser;
+    }
+
+    if (agg_pk != NULL) {
+        if (!secp256k1_xonly_pubkey_serialize(ctx, pk_ser, agg_pk)) {
+            return 0;
+        }
+        pk_ser_ptr = pk_ser;
+    }
+    secp256k1_nonce_function_frost(k, session_id32, msg32, sk_ser_ptr, pk_ser_ptr, extra_input32);
+    VERIFY_CHECK(!secp256k1_scalar_is_zero(&k[0]));
+    VERIFY_CHECK(!secp256k1_scalar_is_zero(&k[1]));
+    VERIFY_CHECK(!secp256k1_scalar_eq(&k[0], &k[1]));
+    secp256k1_frost_secnonce_save(secnonce, k);
+    secp256k1_frost_secnonce_invalidate(ctx, secnonce, !ret);
+
+    for (i = 0; i < 2; i++) {
+        secp256k1_gej nonce_ptj;
+        secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &nonce_ptj, &k[i]);
+        secp256k1_ge_set_gej(&nonce_pt[i], &nonce_ptj);
+        secp256k1_declassify(ctx, &nonce_pt[i], sizeof(nonce_pt));
+        secp256k1_scalar_clear(&k[i]);
+    }
+    /* nonce_pt won't be infinity because k != 0 with overwhelming probability */
+    secp256k1_frost_pubnonce_save(pubnonce, nonce_pt);
+    return ret;
+}
+
+static int secp256k1_frost_sum_nonces(const secp256k1_context* ctx, secp256k1_gej *summed_nonces, const secp256k1_frost_pubnonce * const* pubnonces, size_t n_pubnonces) {
+    size_t i;
+    int j;
+
+    secp256k1_gej_set_infinity(&summed_nonces[0]);
+    secp256k1_gej_set_infinity(&summed_nonces[1]);
+
+    for (i = 0; i < n_pubnonces; i++) {
+        secp256k1_ge nonce_pt[2];
+        if (!secp256k1_frost_pubnonce_load(ctx, nonce_pt, pubnonces[i])) {
+            return 0;
+        }
+        for (j = 0; j < 2; j++) {
+            secp256k1_gej_add_ge_var(&summed_nonces[j], &summed_nonces[j], &nonce_pt[j], NULL);
+        }
+    }
+    return 1;
+}
+
+/* Implements binding factor from draft-irtf-cfrg-frost-04, section 4.4. */
+/* TODO: consider updating to frost-08 to address maleability at the cost of performance */
+/* See https://github.com/cfrg/draft-irtf-cfrg-frost/pull/217 */
+static int secp256k1_frost_compute_noncehash(const secp256k1_context* ctx, unsigned char *noncehash, const unsigned char *msg, const secp256k1_frost_pubnonce * const* pubnonces, size_t n_pubnonces) {
+    unsigned char buf[66];
+    secp256k1_sha256 sha;
+    size_t i;
+
+    secp256k1_sha256_initialize_tagged(&sha, (unsigned char*)"FROST/noncecoef", sizeof("FROST/noncecoef") - 1);
+    /* TODO: sort by index */
+    for (i = 0; i < n_pubnonces; i++) {
+        if (!secp256k1_frost_pubnonce_serialize(ctx, buf, pubnonces[i])) {
+            return 0;
+        }
+        secp256k1_sha256_write(&sha, buf, sizeof(buf));
+    }
+    secp256k1_sha256_write(&sha, msg, 32);
+    secp256k1_sha256_finalize(&sha, noncehash);
+    return 1;
+}
```

Comment:
`frost.md` says

> the API supports an alternative protocol flow where generating the key (see Key Generation above) is allowed to happen after exchanging nonces (step 3).

While this is possible with MuSig2, I haven't considered whether it's possible in FROST. If it is possible in principle, then I believe the `agg_pk` must also be input to the `noncehash` - similar to how this is done in MuSig2.

The security proof for MuSig relies on the following property: For any two executions of the `sign` algorithm, if the challenge `c` in the first execution differs from the challenge `c'` in the second execution, then the nonce coefficient in the first execution `b` must differ from the nonce coefficient in the second execution `b'`. In short: `c != c' => b != b'`.

If `agg_pk` is not hashed into `b` as is the case right now with this PR, then `agg_pk` affects `c` but not `b`, and the above implication does not hold.

Reply:
Great point, I will add `agg_pk` to the `noncehash`.